tests/unit/common/url_decode_tests.cpp

```diff
@@ -0,0 +1,48 @@
+#include "catch2/catch.hpp"
+#include "http/url_decode.h"
+
+#include <string>
+#include <cstring>
+
+using http::url_decode;
+
+TEST_CASE("url decode", "[url]") {
+    std::string common_url = "/some/random/test/url";
+
+    SECTION("decoding weird chars") {
+        std::string url = "/api/v1/files/%21%40%23%24%25%5E%26%2A%28%29%5B%5D%20%7B%7D";
+        char decoded_url[url.size()];
+        CHECK(url_decode(url, decoded_url, sizeof(decoded_url)) == true);
+        CHECK(strcmp(decoded_url, "/api/v1/files/!@#$%^&*()[] {}") == 0);
+    }
+
+    SECTION("not a number after %") {
+        std::string url = "/api/%^7e)";
+        char decoded_url[100];
+        CHECK(url_decode(url, decoded_url, sizeof(decoded_url)) == false);
+        CHECK(strcmp(decoded_url, "/api/") == 0);
+    }
+
+    SECTION("code outside of ascii range") {
+        std::string url = "%ee";
+        char decoded_url[100];
+        CHECK(url_decode(url, decoded_url, sizeof(decoded_url)) == false);
+    }
+
+    SECTION("short buffer") {
+        char decoded_url[1];
+        CHECK(url_decode(common_url, decoded_url, sizeof(decoded_url)) == false);
+    }
+
+    SECTION("zero length buffer") {
+        char decoded_url[0];
+        CHECK(url_decode(common_url, decoded_url, sizeof(decoded_url)) == false);
+    }
+
+    SECTION("empty url") {
+        std::string url = "";
+        char decoded_url[100];
+        CHECK(url_decode(url, decoded_url, sizeof(decoded_url)) == true);
+        CHECK(strcmp(decoded_url, "") == 0);
+    }
```

Comment:
I'd propose few more tests:

* Something with an incomplete escape sequence at the end, eg `/x/a%2`.
* Some integration with the Parser structure ‒ combination with real `?` that should end the path vs. encoded `?` that should be part of the resulting `?`. Similarly, we probably need to make sure one can't overcome the restrictions not to contain `..` in the path by escaping it. That is, we still don't want to allow someone to escape from `/usb` by encoding the dots or slashes into percents.